rspace/src/main/scala/coop/rchain/rspace/history/package.scala

```diff
@@ -201,27 +256,115 @@ package object history {
+                  val newPtr = if (pathLeft.size > 0) {
+                    setupSkipNode(store, txn, LeafPointer(newLeafHash), pathLeft)
+                  } else {
+                    LeafPointer(newLeafHash)
+                  }
+                  Node(
+                    PointerBlock
+                      .create()
+                      .updated(List((oldLeafIndex, ptr), (newLeafIndex, newPtr))))
+                case (ol, nl) if (ol == nl) && (ol > 1) =>
+                  val oldLeafIndex = JByte.toUnsignedInt(oldSuffix(0))
+                  val newLeafIndex = JByte.toUnsignedInt(newSuffix(0))
+                  val newNode = setupSkipNode(store,
+                                              txn,
+                                              LeafPointer(newLeafHash),
+                                              newSuffix.drop(1) ++ pathLeft)
+                  val oldNode = setupSkipNode(store, txn, ptr, oldSuffix.drop(1))
+                  Node(
+                    PointerBlock
+                      .create()
+                      .updated(List((oldLeafIndex, oldNode), (newLeafIndex, newNode))))
+                case _ => throw new InsertException("Affixes have different lengths")
+              }
+
+              val pnHash = Trie.hash(pn)(codecK, codecV)
+              store.put(txn, pnHash, pn)
+
+              val add = if (sharedSubpathLength > 0) {
+                Skip(ByteVector(sharedPrefix), NodePointer(pnHash))
+              } else {
+                pn
+              }
+              //analyze:
+              val rehashedNodes = rehash[K, V](add, parents)
+              val newRootHash   = insertTries(store, txn, rehashedNodes).get
+              store.putRoot(txn, branch, newRootHash)
+              logger.debug(s"workingRootHash: $newRootHash")
           }
       }
     }
 
   @tailrec
-  private[this] def propagateLeafUpward[T, K, V](
-      hash: Blake2b256Hash,
-      parents: Seq[(Int, Trie[K, V])]): (Node, Seq[(Int, Trie[K, V])]) =
+  private[this] def propagateTrieUpward[T, K, V](store: ITrieStore[T, K, V],
+                                                 txn: T,
+                                                 ptr: NonEmptyPointer,
+                                                 incomingAffix: ByteVector,
+                                                 parents: Parents[K, V])(
+      implicit codecK: Codec[K],
+      codecV: Codec[V]): (Trie[K, V], Parents[K, V]) =
     parents match {
       // If the list parents only contains a single Node, we know we are at the root, and we
-      // can update the Vector at the given index to point to the Leaf.
+      // can update the Vector at the given index to point to the node.
       case Seq((byte, Node(pointerBlock))) =>
-        (Node(pointerBlock.updated(List((byte, LeafPointer(hash))))), Seq.empty[(Int, Node)])
+        ptr match {
+          case _: LeafPointer =>
+            //TODO assumption - key > 2. bad assumption
```

Comment:
Help me understand this comment.

Reply:
I would be curious to know what else blows up with arbitrary length keys, so we can make some informed decisions about supporting them.  In the meantime, seems like we should use an `assert` to express this invariant.